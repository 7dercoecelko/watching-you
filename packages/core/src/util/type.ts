function isHtmlElement(x: unknown): x is HTMLElement {
  return x instanceof HTMLElement;
}
export { isHtmlElement };
