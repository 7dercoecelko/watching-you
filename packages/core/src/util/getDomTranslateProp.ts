// 有優化空間
function getDomTranslateProp(dom: HTMLElement) {
  const domTransform = getComputedStyle(dom).transform;
  if (domTransform === 'none') return { x: 0, y: 0 };
  const result = getComputedStyle(dom)
    .transform.split('(')
    .slice(1)[0]
    .slice(0, -1)
    .split(', ')
    .slice(4)
    .map(Number);
  return { x: result[0], y: result[1] };
}

export default getDomTranslateProp;
