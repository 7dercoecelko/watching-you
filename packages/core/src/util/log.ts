type logLevel = 'error' | 'warn' | 'verbose';

function log(message: string, level: logLevel) {
  if (level === 'error') {
    throw Error(message);
  }
  if (process.env.NODE_ENV === 'production') return;
  if (level === 'warn') {
    console.warn(message);
    return;
  }
  if (level === 'verbose') {
    console.log(message);
    return;
  }
}

export default log;
