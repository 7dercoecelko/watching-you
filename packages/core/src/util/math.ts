function round(number: number) {
  return Math.round(number * 10) / 10;
}

export { round };
