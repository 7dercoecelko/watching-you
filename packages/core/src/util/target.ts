import {
  WatchingYouTargetDom,
  WatchingYouTargetInput,
  WatchingYouTargetMouse,
} from '../type/Target';

function isTargetMouse(targetProps?: {
  target?: unknown;
  targetType?: unknown;
}): targetProps is WatchingYouTargetMouse {
  return (
    !targetProps ||
    targetProps.targetType === 'mouse' ||
    (!targetProps.targetType && !targetProps.target)
  );
}
function isTargetDom(targetProps?: {
  target?: unknown;
  targetType?: unknown;
}): targetProps is WatchingYouTargetDom {
  if (!targetProps) return false;
  return (
    targetProps.targetType === 'dom' ||
    (!targetProps.targetType && !!targetProps.target)
  );
}
function isTargetInput(targetProps?: {
  target?: unknown;
  targetType?: unknown;
}): targetProps is WatchingYouTargetInput {
  if (!targetProps) return false;
  // if(!watch) log(`If the targetType parameter is input, then the watch must be a input`,'warn');
  return targetProps.targetType === 'input';
}

export { isTargetMouse, isTargetDom, isTargetInput };
