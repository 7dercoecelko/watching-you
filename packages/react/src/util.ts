function concatObject(...a: unknown[]) {
  return Object.assign({}, ...a);
}

export { concatObject };
