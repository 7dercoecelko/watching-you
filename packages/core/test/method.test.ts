/**
 * @vitest-environment happy-dom
 */
import { beforeEach, describe, expect, it } from 'vitest';
import WatchingYou from '../src/index';
import { mousemove, wait } from './util';

describe('Method', () => {
  beforeEach(() => {
    document.body.innerHTML = `
      <div class="eyes">
        <div>O</div>
      </div>
    `;
  });

  it('It should start watching something, when the start method is call', async () => {
    const eyesDom = document.querySelector<HTMLElement>('.eyes');
    if (eyesDom === null) throw Error('eyesDom is not found!');
    const watchingYou = new WatchingYou(eyesDom);
    expect(eyesDom.style.transform).toBe('');
    watchingYou.start();
    await mousemove(100, 200);
    expect(eyesDom.style.transform).not.toBe('');
    expect(eyesDom.style.transform).not.toBe(
      'translate3d(0px,0px,0px) rotate(0deg)',
    );
    watchingYou.stop();
  });

  it('It should go back to the original point, when the stop method is call', async () => {
    const eyesDom = document.querySelector<HTMLElement>('.eyes');
    if (eyesDom === null) throw Error('eyesDom is not found!');
    const watchingYou = new WatchingYou(eyesDom);
    watchingYou.start();
    await mousemove(100, 200);
    expect(eyesDom.style.transform).not.toBe('');
    expect(eyesDom.style.transform).not.toBe(
      'translate3d(0px,0px,0px) rotate(0deg)',
    );
    watchingYou.stop();
    expect(eyesDom.style.transform).toBe(
      'translate3d(0px,0px,0px) rotate(0deg)',
    );
  });

  it('Its coordinates should be changed when the power is changed', async () => {
    const eyesDom = document.querySelector<HTMLElement>('.eyes');
    if (eyesDom === null) throw Error('eyesDom is not found!');
    const watchingYou = new WatchingYou(eyesDom);
    watchingYou.start();
    watchingYou.setPower(100);
    await mousemove(500, 500);
    expect(eyesDom.style.transform).toMatchSnapshot();
    watchingYou.setPower(10);
    await mousemove(501, 501);
    expect(eyesDom.style.transform).toMatchSnapshot();
    watchingYou.setPower({ x: 100, y: 0 });
    await mousemove(502, 502);
    expect(eyesDom.style.transform).toMatchSnapshot();
  });
});
