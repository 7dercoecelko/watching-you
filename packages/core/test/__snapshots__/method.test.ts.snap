// Vitest Snapshot v1

exports[`Method > Its coordinates should be changed when the power is changed 1`] = `"translate3d(75px,66.2px,0px) rotate(131.4deg)"`;

exports[`Method > Its coordinates should be changed when the power is changed 2`] = `"translate3d(7.1px,7px,0px) rotate(134.6deg)"`;

exports[`Method > Its coordinates should be changed when the power is changed 3`] = `"translate3d(70.7px,0px,0px) rotate(135deg)"`;
